crates/subspace-networking/src/peer_info.rs

```diff
@@ -0,0 +1,207 @@
+    /// Creates a new peer [`ConstantPeerInfoProvider`].
+    pub fn new(peer_info: PeerInfo) -> Self {
+        Self { peer_info }
+    }
+}
+
+impl PeerInfoProvider for ConstantPeerInfoProvider {
+    fn peer_info(&self) -> PeerInfo {
+        self.peer_info.clone()
+    }
+}
+
+/// Event generated by the `Peer Info` network behaviour.
+#[derive(Debug)]
+pub struct Event {
+    /// The peer ID of the remote.
+    pub peer_id: PeerId,
+    /// The result of an inbound or outbound peer info request.
+    pub result: Result,
+}
+
+impl<PIP: PeerInfoProvider> Behaviour<PIP> {
+    /// Creates a new `Peer Info` network behaviour with the given configuration.
+    pub fn new(config: Config, peer_info_provider: PIP) -> Self {
+        Self {
+            config,
+            peer_info_provider,
+            events: VecDeque::new(),
+            requests: Vec::new(),
+        }
+    }
+}
+
+impl<PIP: PeerInfoProvider> NetworkBehaviour for Behaviour<PIP> {
+    type ConnectionHandler = Handler;
+    type OutEvent = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> std::result::Result<THandler<Self>, ConnectionDenied> {
+        Ok(Handler::new(self.config.clone()))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+    ) -> std::result::Result<THandler<Self>, ConnectionDenied> {
+        Ok(Handler::new(self.config.clone()))
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer: PeerId,
+        _: ConnectionId,
+        result: THandlerOutEvent<Self>,
+    ) {
+        self.events.push_front(Event {
+            peer_id: peer,
+            result,
+        })
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut Context<'_>,
+        _: &mut impl PollParameters,
+    ) -> Poll<ToSwarm<Self::OutEvent, THandlerInEvent<Self>>> {
+        if let Some(e) = self.events.pop_back() {
+            let Event { result, peer_id } = &e;
+
+            match result {
+                Ok(PeerInfoSuccess::DataSent) => {
+                    debug!(%peer_id,"Peer info sent.", )
+                }
+                Ok(PeerInfoSuccess::Received(_)) => {
+                    debug!(%peer_id, "Peer info received", )
+                }
+                Err(err) => {
+                    debug!(%peer_id, ?err, "Peer info error");
+                }
+            }
+
+            return Poll::Ready(ToSwarm::GenerateEvent(e));
+        }
+
+        // Check for pending requests.
+        if let Some(Request { peer_id }) = self.requests.pop() {
+            return Poll::Ready(ToSwarm::NotifyHandler {
+                peer_id,
+                handler: NotifyHandler::Any,
+                event: HandlerInEvent {
+                    peer_info: self.peer_info_provider.peer_info(),
```

Comment:
`.peer_info()` does expensive locking and serialization that will be exactly the same for everyone, we should ideally call it once and store, maybe in an `Arc` in `Request` itself, so we don't serialize the same thing over and over again for every peer.